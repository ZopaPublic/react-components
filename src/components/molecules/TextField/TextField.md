### Summary

`<TextField />` is a convenient wrapper composed of four different components:

- [`<SizedContainer />`](/#/Components/Layout/SizedContainer)
- [`<InputText />`](/#/Components/Atoms/InputText)
- [`<InputLabel />`](/#/Components/Atoms/InputLabel)
- [`<ErrorMessage />`](/#/Components/Atoms/ErrorMessage)

⚠️ &nbsp;Note that a `name` prop **must be provided** to automatically set:

- `htmlFor` prop on `<InputLabel />`
- `data-automation` prop on `<ErrorMessage />`
- `id` prop on `<InputText />`

### Examples

- Default

```tsx
import { TextField } from '@zopauk/react-components';

<TextField name="text1" />;
```

- With a specific size

```tsx
import { TextField } from '@zopauk/react-components';

<TextField inputSize="short" name="text2" />;
```

- With a label

```tsx
import { TextField } from '@zopauk/react-components';

<TextField label="First name" name="text3" />;
```

- With an icon

```tsx
import { TextField, Icon, colors } from '@zopauk/react-components';
import { faCoffee } from '@fortawesome/free-solid-svg-icons';

<TextField startIcon={<Icon variant={faCoffee} />} label="First name" name="text4" />;
```

- With error message

```tsx
import { TextField } from '@zopauk/react-components';

<TextField errorMessage="Oops ! Error !" name="text5" />;
```

- Valid state

```tsx
import { TextField } from '@zopauk/react-components';

<TextField isValid={true} name="text6" />;
```

- With all the previous configuration

```tsx
import { TextField, Icon, colors } from '@zopauk/react-components';
import { faCoffee } from '@fortawesome/free-solid-svg-icons';

<TextField
  label="Label"
  startIcon={<Icon variant={faCoffee} />}
  helpText="Additional info"
  errorMessage="Not ok!"
  inputSize="short"
  name="text7"
/>;
```
