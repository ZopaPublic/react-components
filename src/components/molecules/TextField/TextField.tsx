import React, { forwardRef } from 'react';
import styled from 'styled-components';
import ErrorMessage from '../../atoms/ErrorMessage/ErrorMessage';
import Text from '../../atoms/Text/Text';
import InputText from '../../atoms/InputText/InputText';
import InputLabel from '../../atoms/InputLabel/InputLabel';
import SizedContainer from '../../layout/SizedContainer/SizedContainer';
import { colors, typography } from '../../../constants';
import { FieldProps, InputProps } from '../../types';

export interface TextFieldProps extends FieldProps, InputProps {
  prefix?: string;
}

export interface PrefixProps {
  prefix: string;
}

export interface LabelProps {
  withHelpText?: boolean;
}
const Label = styled(InputLabel)<LabelProps>`
  ${({ withHelpText }) => withHelpText && `margin-bottom: 0;`}
`;

const FieldError = styled(ErrorMessage)`
  margin-top: 5px;
`;

const HelpText = styled(Text).attrs({
  size: 'small',
})`
  margin-bottom: 5px;
  display: block;
`;

const Prefix = styled.span<PrefixProps>`
  position: relative;
  display: block;
  
  &::before {
    content: '${({ prefix }: PrefixProps) => prefix}';
    position: absolute;
    left: 10px;
    top: 0;
    bottom: 0;
    display: flex;
    align-items: center;
    font-size: ${typography.sizes.text.body};
    color: ${colors.greyDarkest};
  }
  
  input {
    padding-left: 24px;
  }
`;

const TextField = forwardRef<HTMLInputElement, TextFieldProps>(
  ({ label, errorMessage, isValid, name, inputSize, helpText, prefix, ...rest }, ref) => {
    if (!name) {
      throw Error('Name must be set in inputProps. Check the docs.');
    }

    if (prefix && prefix.length > 1) {
      throw Error('Prefixes can only have one character');
    }

    const input = (
      <InputText
        id={`text-id-${name}`}
        type="text"
        hasError={!!errorMessage}
        isValid={isValid}
        aria-label={label ? undefined : name}
        ref={ref}
        name={name}
        {...rest}
      />
    );

    return (
      <>
        {label && (
          <Label withHelpText={!!helpText} htmlFor={`text-id-${name}`}>
            {label}
          </Label>
        )}
        {helpText && <HelpText size="small">{helpText}</HelpText>}
        <SizedContainer size={inputSize}>{prefix ? <Prefix prefix={prefix}>{input}</Prefix> : input}</SizedContainer>
        {errorMessage && <FieldError data-automation={`ZA.error-${name}`}>{errorMessage}</FieldError>}
      </>
    );
  },
);

export default TextField;
