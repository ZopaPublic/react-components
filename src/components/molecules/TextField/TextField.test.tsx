import React from 'react';
import { render } from '@testing-library/react';
import axe from '../../../../axe-helper';
import TextField from './TextField';

describe('<TextField />', () => {
  it('renders the component with no a11y violations', async () => {
    const { container } = render(<TextField name="test1" />);
    const { getComputedStyle } = window;
    // Fix warning in console
    window.getComputedStyle = (elt) => getComputedStyle(elt);
    const results = await axe(container.innerHTML);

    expect(container.firstChild).toMatchSnapshot();
    expect(results).toHaveNoViolations();
  });

  it('renders the component with label', () => {
    const { container } = render(<TextField name="test1" label="Username" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders the component and isValid set to true', () => {
    const { container } = render(<TextField name="test1" isValid={true} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders the component with error', () => {
    const { container } = render(<TextField name="test1" errorMessage="Ops !" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders the component with size', () => {
    const { container } = render(<TextField name="test1" inputSize="short" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders the component with error and label and size', () => {
    const { container } = render(
      <TextField name="test1" helpText="Additional info" errorMessage="Ops !" label="yes!" inputSize="long" />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders the component with prefix', () => {
    const { container } = render(<TextField name="test1" inputSize="short" prefix="Â£" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('throws if no name provided', () => {
    console.error = jest.fn();
    expect(() => render(<TextField />)).toThrow('Name must be set in inputProps. Check the docs.');
  });

  it('throws if prefix is longer than one character', () => {
    console.error = jest.fn();
    expect(() => render(<TextField name="test1" prefix="$$" />)).toThrow('Prefixes can only have one character');
  });
});
