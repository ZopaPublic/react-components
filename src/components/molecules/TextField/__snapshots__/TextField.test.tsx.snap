// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> renders the component and isValid set to true 1`] = `
.c1 {
  position: relative;
}

.c2 {
  width: 100%;
  -webkit-appearance: none;
  outline: none;
  border-radius: 8px;
  height: 50px;
  padding: 0 16px;
  font-size: 16px;
  font-weight: 400;
  color: #4A545E;
  border: 1px solid #3EBC64;
  box-shadow: 0 0 4px 0 transparent;
  -webkit-transition-property: border,box-shadow;
  transition-property: border,box-shadow;
  -webkit-transition: 0.2s ease-in-out;
  transition: 0.2s ease-in-out;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
}

.c2:hover {
  border: 1px solid #009E8E;
  box-shadow: 0 0 4px 0 #009E8E;
}

.c2:focus {
  border: 1px solid #009E8E;
  box-shadow: 0 0 4px 0 #009E8E #009E8E;
}

.c2::-webkit-input-placeholder {
  color: #D4D7D9;
  opacity: 1;
}

.c2::-moz-placeholder {
  color: #D4D7D9;
  opacity: 1;
}

.c2:-ms-input-placeholder {
  color: #D4D7D9;
  opacity: 1;
}

.c2::placeholder {
  color: #D4D7D9;
  opacity: 1;
}

.c2:disabled {
  -webkit-text-fill-color: #818F9B;
  opacity: 1;
  border: 1px solid #3EBC64;
  box-shadow: 0 0 4px 0 transparent;
  background-color: #F7F7F7;
  cursor: not-allowed;
}

.c0 {
  max-width: 100%;
  width: 100%;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <input
      aria-label="test1"
      class="c2"
      id="text-id-test1"
      name="test1"
      type="text"
    />
  </div>
</div>
`;

exports[`<TextField /> renders the component with error 1`] = `
.c1 {
  position: relative;
}

.c2 {
  width: 100%;
  -webkit-appearance: none;
  outline: none;
  border-radius: 8px;
  height: 50px;
  padding: 0 16px;
  font-size: 16px;
  font-weight: 400;
  color: #4A545E;
  border: 1px solid #FF4539;
  box-shadow: 0 0 4px 0 transparent;
  -webkit-transition-property: border,box-shadow;
  transition-property: border,box-shadow;
  -webkit-transition: 0.2s ease-in-out;
  transition: 0.2s ease-in-out;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
}

.c2:hover {
  border: 1px solid #009E8E;
  box-shadow: 0 0 4px 0 #009E8E;
}

.c2:focus {
  border: 1px solid #009E8E;
  box-shadow: 0 0 4px 0 #009E8E #009E8E;
}

.c2::-webkit-input-placeholder {
  color: #D4D7D9;
  opacity: 1;
}

.c2::-moz-placeholder {
  color: #D4D7D9;
  opacity: 1;
}

.c2:-ms-input-placeholder {
  color: #D4D7D9;
  opacity: 1;
}

.c2::placeholder {
  color: #D4D7D9;
  opacity: 1;
}

.c2:disabled {
  -webkit-text-fill-color: #818F9B;
  opacity: 1;
  border: 1px solid #FF4539;
  box-shadow: 0 0 4px 0 transparent;
  background-color: #F7F7F7;
  cursor: not-allowed;
}

.c0 {
  max-width: 100%;
  width: 100%;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <input
      aria-label="test1"
      class="c2"
      id="text-id-test1"
      name="test1"
      type="text"
    />
  </div>
</div>
`;

exports[`<TextField /> renders the component with error and label and size 1`] = `
.c0 {
  display: block;
  margin: 0 0 10px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  color: #2C3236;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  font-weight: 600;
  font-size: 16px;
}

.c1 {
  margin-bottom: 0;
}

<label
  class="c0 c1"
  for="text-id-test1"
>
  yes!
</label>
`;

exports[`<TextField /> renders the component with label 1`] = `
.c0 {
  display: block;
  margin: 0 0 10px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  color: #2C3236;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  font-weight: 600;
  font-size: 16px;
}

<label
  class="c0 "
  for="text-id-test1"
>
  Username
</label>
`;

exports[`<TextField /> renders the component with no a11y violations 1`] = `
.c1 {
  position: relative;
}

.c2 {
  width: 100%;
  -webkit-appearance: none;
  outline: none;
  border-radius: 8px;
  height: 50px;
  padding: 0 16px;
  font-size: 16px;
  font-weight: 400;
  color: #4A545E;
  border: 1px solid #818F9B;
  box-shadow: 0 0 4px 0 transparent;
  -webkit-transition-property: border,box-shadow;
  transition-property: border,box-shadow;
  -webkit-transition: 0.2s ease-in-out;
  transition: 0.2s ease-in-out;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
}

.c2:hover {
  border: 1px solid #009E8E;
  box-shadow: 0 0 4px 0 #009E8E;
}

.c2:focus {
  border: 1px solid #009E8E;
  box-shadow: 0 0 4px 0 #009E8E #009E8E;
}

.c2::-webkit-input-placeholder {
  color: #D4D7D9;
  opacity: 1;
}

.c2::-moz-placeholder {
  color: #D4D7D9;
  opacity: 1;
}

.c2:-ms-input-placeholder {
  color: #D4D7D9;
  opacity: 1;
}

.c2::placeholder {
  color: #D4D7D9;
  opacity: 1;
}

.c2:disabled {
  -webkit-text-fill-color: #818F9B;
  opacity: 1;
  border: 1px solid #818F9B;
  box-shadow: 0 0 4px 0 transparent;
  background-color: #F7F7F7;
  cursor: not-allowed;
}

.c0 {
  max-width: 100%;
  width: 100%;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <input
      aria-label="test1"
      class="c2"
      id="text-id-test1"
      name="test1"
      type="text"
    />
  </div>
</div>
`;

exports[`<TextField /> renders the component with prefix 1`] = `
.c2 {
  position: relative;
}

.c3 {
  width: 100%;
  -webkit-appearance: none;
  outline: none;
  border-radius: 8px;
  height: 50px;
  padding: 0 16px;
  font-size: 16px;
  font-weight: 400;
  color: #4A545E;
  border: 1px solid #818F9B;
  box-shadow: 0 0 4px 0 transparent;
  -webkit-transition-property: border,box-shadow;
  transition-property: border,box-shadow;
  -webkit-transition: 0.2s ease-in-out;
  transition: 0.2s ease-in-out;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
}

.c3:hover {
  border: 1px solid #009E8E;
  box-shadow: 0 0 4px 0 #009E8E;
}

.c3:focus {
  border: 1px solid #009E8E;
  box-shadow: 0 0 4px 0 #009E8E #009E8E;
}

.c3::-webkit-input-placeholder {
  color: #D4D7D9;
  opacity: 1;
}

.c3::-moz-placeholder {
  color: #D4D7D9;
  opacity: 1;
}

.c3:-ms-input-placeholder {
  color: #D4D7D9;
  opacity: 1;
}

.c3::placeholder {
  color: #D4D7D9;
  opacity: 1;
}

.c3:disabled {
  -webkit-text-fill-color: #818F9B;
  opacity: 1;
  border: 1px solid #818F9B;
  box-shadow: 0 0 4px 0 transparent;
  background-color: #F7F7F7;
  cursor: not-allowed;
}

.c0 {
  max-width: 148px;
  width: 100%;
}

.c1 {
  position: relative;
  display: block;
}

.c1::before {
  content: '£';
  position: absolute;
  left: 10px;
  top: 0;
  bottom: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 16px;
  color: #2C3236;
}

.c1 input {
  padding-left: 24px;
}

<div
  class="c0"
>
  <span
    class="c1"
    prefix="£"
  >
    <div
      class="c2"
    >
      <input
        aria-label="test1"
        class="c3"
        id="text-id-test1"
        name="test1"
        type="text"
      />
    </div>
  </span>
</div>
`;

exports[`<TextField /> renders the component with size 1`] = `
.c1 {
  position: relative;
}

.c2 {
  width: 100%;
  -webkit-appearance: none;
  outline: none;
  border-radius: 8px;
  height: 50px;
  padding: 0 16px;
  font-size: 16px;
  font-weight: 400;
  color: #4A545E;
  border: 1px solid #818F9B;
  box-shadow: 0 0 4px 0 transparent;
  -webkit-transition-property: border,box-shadow;
  transition-property: border,box-shadow;
  -webkit-transition: 0.2s ease-in-out;
  transition: 0.2s ease-in-out;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
}

.c2:hover {
  border: 1px solid #009E8E;
  box-shadow: 0 0 4px 0 #009E8E;
}

.c2:focus {
  border: 1px solid #009E8E;
  box-shadow: 0 0 4px 0 #009E8E #009E8E;
}

.c2::-webkit-input-placeholder {
  color: #D4D7D9;
  opacity: 1;
}

.c2::-moz-placeholder {
  color: #D4D7D9;
  opacity: 1;
}

.c2:-ms-input-placeholder {
  color: #D4D7D9;
  opacity: 1;
}

.c2::placeholder {
  color: #D4D7D9;
  opacity: 1;
}

.c2:disabled {
  -webkit-text-fill-color: #818F9B;
  opacity: 1;
  border: 1px solid #818F9B;
  box-shadow: 0 0 4px 0 transparent;
  background-color: #F7F7F7;
  cursor: not-allowed;
}

.c0 {
  max-width: 148px;
  width: 100%;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <input
      aria-label="test1"
      class="c2"
      id="text-id-test1"
      name="test1"
      type="text"
    />
  </div>
</div>
`;
